optuna/samplers/nsga2/_crossovers/_base.py

```diff
@@ -0,0 +1,58 @@
+import abc
+from typing import Dict
+
+import numpy as np
+
+from optuna.distributions import BaseDistribution
+from optuna.study import Study
+
+
+class BaseCrossover(object, metaclass=abc.ABCMeta):
+    """Base class for crossovers.
+
+    A crossover operation is used by :class:`~optuna.samplers.NSGAIISampler`
+    to create new parameter combination from parameters of ``n`` parent individuals.
+    """
+
+    def __str__(self) -> str:
+
+        return self.__class__.__name__
+
+    @property
+    @abc.abstractmethod
+    def n_parents(self) -> int:
+        """Number of parent individuals required to perform crossover."""
+
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def crossover(
+        self,
+        parents_params: np.ndarray,
+        rng: np.random.RandomState,
+        study: Study,
+        search_space: Dict[str, BaseDistribution],
```

Comment:
Just an idea, but we may be able to define a function converting `__SearchSpaceTransform.bounds` to `List[BaseDistribution]` and apply the function to `search_space` before passing to the `crossover` method.

Reply:
Since `_SearchSpaceTransform` already has an information about search space and bounds, maybe it should be able to do this conversion? I could implement this method in separate PR and merge it back here.

Reply:
Or not, since it might not be possible to instantiate a distribution with transformed bounds e.g. `FloatDistribution(log(1), log(10), log=True)`.

Reply:
@hvy As the implementor of `_SearchSpaceTransform`, do you have any preference for how to deal with this issue?

Reply:
Thanks for mentioning! I'm still going through the changes so might be off, but let me just check the following.

> A difficult point is that, in categorical distributions case, we should keep using Dict[str, BaseDistrivution] because we don't apply the transformation on categorical parameters.

This sentence. Does this refer exclusively to the `UniformCrossover` being used for the categorical parameters in `_try_crossover`? If so, I'm wondering if we cannot design `BaseCrossover.crossover` in such a way that only transformed and numerical parameters are expected. If necessary, implement the uniform categorical crossover logic inlined in `_try_crossover`. This along with changing the argument to `crossover` to the transform object instead of the distribution dict.

(This is off topic, but I'm also curious if we cannot transform (one-hot encode) categorical parameters and perform crossover on them as well (with arbitrary crossover algorithm). )

Reply:
> Does this refer exclusively to the `UniformCrossover` being used for the categorical parameters in `_try_crossover`? 

Yes.

>  I'm wondering if we cannot design `BaseCrossover.crossover` in such a way that only transformed and numerical parameters are expected. 

Sounds good. What do you think of that? > @xadrianzetx @HideakiImamura 

>  I'm also curious if we cannot transform (one-hot encode) categorical parameters and perform crossover on them as well

This could be the best choice if we can. However, to decide whether this approach is acceptable or not, it would be needed to run a comprehensive benchmark as in #2903. So, it feels not an easy task and out of the scope of this PR, I think.

Reply:
> This could be the best choice if we can. However, to decide whether this approach is acceptable or not, it would be needed to run a comprehensive benchmark as in https://github.com/optuna/optuna/pull/2903. So, it feels not an easy task and out of the scope of this PR, I think.

+1

Reply:
> I'm wondering if we cannot design BaseCrossover.crossover in such a way that only transformed and numerical parameters are expected.

This could be assumed about current design, if categorical parameters would be numerical encoded instead of OHE to fit `parents_params` arg. If we are to keep `parents_param` as numpy array, I think shaping it as `num_parents x num_params` is reasonable, but there's no way to fit one hot encoded params there.

IMO the main problem is still that we are trying to pass `search_space` dict, which might contain outdated information about distribution bounds by the time `crossover` is called. I'm not sure about passing `_SearchSpaceTransform` instance instead, as this is a private API and I designed all crossover APIs as public since  I'm assuming users should be able to define their own (same as eg. with storage). This would put user in position where they need to `from optuna._transform import _SearchSpaceTransform`.

Since current use of `search_space` within crossover classes is limited to `len()` and bounds check, I think @hvy approach of inlining uniform crossover for categorical params is reasonable, and all `BaseCrossover` implementations should take `search_space` in form of `_SearchSpaceTransform.bounds` (numpy array) and only deal with numerical params. If there is ever a need to have param names or other distribution related attributes (like `log` or `step`), then we can propably wrap this bounds array with object that holds more info about distributions after transformation.

Reply:
> I designed all crossover APIs as public

Ah, thank you. This is an important aspect. If bounds suffice for now, I agree that it is simpler to just pass the those as you say.

The API would be less Optuna-like (NumPy `ndarrays` over of `Study` or `FrozenTrial`/`Trial`), but that might be fine, somewhat similar to the BoTorch sampler which passes around PyTorch tensors https://github.com/optuna/optuna/blob/master/optuna/integration/botorch.py. I'd however if possible reserve ourselves for feature changes. I.e. we could make it more Optuna-like by passing arguments that more resemble those of `_try_crossover` (e.f. `FrozenTrial`s), along with making `_SearchSpaceTransform` public.

Reply:
> along with making _SearchSpaceTransform public

I think I could work on this right now in separate PR. This would be really usefull as now public `SearchSpaceTransform` would have essentially the functionality of `bounds` wrapper I was talking about previously. I think in the process of making it public we could think about some convenience methods such as `__len__` to return search space length. What do you think about discussing and implementing it in separate PR? It would be a drop in replacement for `search_space` dict.

> we could make it more Optuna-like by passing arguments that more resemble those of _try_crossover

Passing some object with params instead of an array in `parents_params`? +1 to that in future revisions, as it would probably come in handy if other crossovers were to support OHE.

Reply:
> What do you think about discussing and implementing it in separate PR?

Sounds good. I think it'll be important to hear what others are thinking about this, considering the fact that this'll likely be an API targeting only a small portion of the users. `__len__` by the way sounds interesting. I'd personally prefer named methods, as it's not entirely clear from `__len__` whether it will return the length after or prior to the transformation, but this is not a strong opinion.

Given the discussion so far, how do you suggest we proceed with this PR. Should we aim to merge it as is, revisit it at a later time? If so, should be perhaps make APIs in this PR experimental to reserve ourselves for changes? Alternatively, we'll keep this PR open until everything discussed so far is fixed.

Reply:
+1 to make `_SearchSpaceTransform` public. I think it would be beneficial for users to make it public, since it could be used by users who want to implement a new sampler, for example.

On the other hand, I don't think merging this PR will have to wait for that. Frankly speaking, @xadrianzetx's contribution to Optuna V3 is very big and we are relying on your contribution a lot. Making `_SearchSpaceTransform` public is out of scope for Optuna V3, but perhaps PR is expected to be somewhat bigger, including the interface. Therefore, in this PR, why not perform an in-line uniform crossover on categorical parameters and pass `_SearchSpaceTransform.bounds` to try_crossover? I suggest that making `_SearchSpaceTransform` public is a separate issue to be made into future work.

Reply:
Fully agree with @HideakiImamura.
* Uniform crossover for categorical params will be inlined, and `BaseCrossover` implementations will be reserved for transformed numerical params (for now).
* We will re-design `search_space` arg to accept numpy array of `_SearchSpaceTransform.bounds` instead of distribution dict to solve issue reported by @sile.
* Making `_SearchSpaceTransform` a public API and passing it instead of just `bounds` will be left for future discussion and work (which I'm happy to be assigned to).
* making crossovers accept one-hot-encoded params will be left for future discussion.

Reply:
Perfect! Thank you very much for this useful discussion! > @xadrianzetx @hvy @HideakiImamura